### DISCUSSION TITLE
Module/Environment Changes for 0.68/0.69

### DISCUSSION THREAD

**Discussion by ehawman** [nushell/nushell#6513] [General]
With the release of [0.68 and its changes to modules/overlays/env-vars](https://www.nushell.sh/blog/2022-09-06-nushell-0_68.html), I must admit I'm a bit lost when it comes to updating my current stable of scripts and functions, and I'm likely not alone. I envision this discussion thread as a general "My script broke after updating to 0.68/0.69" knowledge repository, where people can freeform post their issues and read over the problems that others have submitted to learn by example (which is the best way I learn). If this isn't in keeping with the spirit of the board, please close this off :)

I'll kick us off.

`oh-my-posh.nu` is generated by [Oh-My-Posh](https://ohmyposh.dev/docs/installation/prompt) via its own command (`oh-my-posh init nu`)

``` nush
let-env POWERLINE_COMMAND = 'oh-my-posh'
let-env POSH_THEME = 'C:\Users\ehawman\AppData\Local\Programs\oh-my-posh\themes\devious-diamonds.omp.yaml'
let-env PROMPT_INDICATOR = ""
# By default displays the right prompt on the first line
# making it annoying when you have a multiline prompt
# making the behavior different compared to other shells
let-env PROMPT_COMMAND_RIGHT = {''}
let-env NU_VERSION = (version | get version)

# PROMPTS
let-env PROMPT_MULTILINE_INDICATOR = (^'C:/Users/ehawman/AppData/Local/Programs/oh-my-posh/bin/oh-my-posh.exe' print secondary $"--config=($env.POSH_THEME)" --shell=nu $"--shell-version=($env.NU_VERSION)")

let-env PROMPT_COMMAND = {
    let width = (term size -c | get columns | into string)
    ^'C:/Users/ehawman/AppData/Local/Programs/oh-my-posh/bin/oh-my-posh.exe' print primary $"--config=($env.POSH_THEME)" --shell=nu $"--shell-version=($env.NU_VERSION)" $"--execution-time=($env.CMD_DURATION_MS)" $"--error=($env.LAST_EXIT_CODE)" $"--terminal-width=($width)"
}
```

Now by my reading this does nothing but set a bevy of env vars, so I wrapped the entire thing in an `export-env`

``` nush
export-env {
    let-env POWERLINE_COMMAND = 'oh-my-posh'
    let-env POSH_THEME = 'C:\Users\ehawman\AppData\Local\Programs\oh-my-posh\themes\devious-diamonds.omp.yaml'
    let-env PROMPT_INDICATOR = ""
    # By default displays the right prompt on the first line
    # making it annoying when you have a multiline prompt
    # making the behavior different compared to other shells
    let-env PROMPT_COMMAND_RIGHT = {''}
    let-env NU_VERSION = (version | get version)

    # PROMPTS
    let-env PROMPT_MULTILINE_INDICATOR = (^'C:/Users/ehawman/AppData/Local/Programs/oh-my-posh/bin/oh-my-posh.exe' print secondary $"--config=($env.POSH_THEME)" --shell=nu $"--shell-version=($env.NU_VERSION)")

    let-env PROMPT_COMMAND = {
        let width = (term size -c | get columns | into string)
        ^'C:/Users/ehawman/AppData/Local/Programs/oh-my-posh/bin/oh-my-posh.exe' print primary $"--config=($env.POSH_THEME)" --shell=nu $"--shell-version=($env.NU_VERSION)" $"--execution-time=($env.CMD_DURATION_MS)" $"--error=($env.LAST_EXIT_CODE)" $"--terminal-width=($width)"
    }
}
```

then put `source-env C:/Users/ehawman/.config/nu/functions/oh-my-posh.nu` in `config.nu`.

However, this is now what I get 

```
thread 'main' panicked at 'internal error: missing block', crates\nu-protocol\src\engine\engine_state.rs:677:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

[process exited with code 101 (0x00000065)]
```

I haven't quite puzzled out how to run with `RUST_BACKTRACE=1` (`let-env RUST_BACKTRACE=1` in `config.nu` gives me a "missing initial value" error and `let-env RUST_BACKTRACE = 1` acts like I've done nothing at all.

Obviously I'm misunderstanding some basic syntax here. Any thoughts?

**Comment by fdncred** [comment]
I wonder if this could be related to https://github.com/nushell/nushell/issues/6510 even though it doesn't appear to be using `source-env`.

**Comment by kubouch** [comment]
Your solution by wrapping the `let-env`s into the `export-env` is correct, but we have a bug in `source-env` that causes the panic, as fdncred pointed out.

**Reply by ehawman** [reply]
Ah good to know. Thanks! Good luck squashing the bug!

**Comment by ehawman** [comment]
To continue my own thread:

``` nush
#!/usr/bin/env nush

def bigwin [param?] {
	let stored = $env.config.disable_table_indexes
	let-env config = ($env.config | upsert disable_table_indexes true)
	echo [["Updating"]; ["Winget"]]
	winget upgrade --all
	echo [["Complete" "Next"]; ["Winget" "Chocolatey"]]
	choco upgrade all -y
	echo [["Complete" "Next"]; ["Chocolatey" "Scoop"]]
	scoop update
	scoop update *
	echo [["Complete"]; ["Scoop"]]
	let-env config = ($env.config | upsert disable_table_indexes $stored)
}
```

I can put `use ~/.config/nu/functions/bigwin.nu` in `config.nu` to no avail.

I can run that command directly with a similar lack of result. I've also tried every combination of `(export|'') (def|def-env)` that I can think of.

**Reply by ehawman** [reply]
Ok I figured it out.

In my example above, what I created was a module (`bigwin`) which contained a function (`bigwin`). To call the function, I'd first need to correctly export it (`export def bigwin [param?] {`), then execute `bigwin bigwin`.

To remove the extra layer of import, I've changed my `use` call to:

`use ~/.config/nu/functions/bigwin.nu *`, which imports all the exported functions inside the module (which in hindsight makes complete logical sense).
